### PR TITLE
Optimize selected rows for dist lookup table with pthread rwlock

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: paddle/fluid/framework/selected_rows.cc [hunk 1]
```diff
@@ -120,66 +120,86 @@ bool SelectedRows::HasKey(int64_t key) const {
...
+  return ret;
+}
+
+int64_t SelectedRows::AutoGrownIndex(int64_t key) {
+  pthread_rwlock_rdlock(&rwlock_.get()->lock);
+  auto iter = id_to_index_.find(key);
+  if (iter == id_to_index_.end()) {
+    pthread_rwlock_unlock(&rwlock_.get()->lock);
+    pthread_rwlock_wrlock(&rwlock_.get()->lock);
+    PADDLE_ENFORCE(id_to_index_.size() == rows_.size(),
```
Comment (seiriosPlus) [inline paddle/fluid/framework/selected_rows.cc]: Maybe write out the size in the message is better.

File: paddle/fluid/framework/selected_rows.cc [hunk 2]
```diff
@@ -120,66 +120,86 @@ bool SelectedRows::HasKey(int64_t key) const {
...
+  if (!has_key) {
+    pthread_rwlock_unlock(&rwlock_.get()->lock);
+    PADDLE_THROW("Id %s not in table", key);
+  }
+  auto ret = iter->second;
+  pthread_rwlock_unlock(&rwlock_.get()->lock);
+  return ret;
+}
+
+int64_t SelectedRows::AutoGrownIndex(int64_t key) {
```
Comment (seiriosPlus) [inline paddle/fluid/framework/selected_rows.cc]: ```AutoGrownIndex ``` and ```AutoIndex ``` can merge? add function parameter ```grown```?

Comment (jacquesqiao) [reply paddle/fluid/framework/selected_rows.cc]: 嗯，当时主要担心各种判断条件太多，我看看能不能写的清晰一点

File: paddle/fluid/framework/selected_rows.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/selected_rows.cc]: done

File: paddle/fluid/framework/selected_rows.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/selected_rows.cc]: done

File: paddle/fluid/framework/rw_lock.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+limitations under the License. */
+
+#pragma once
+
+#include <pthread.h>
+
+namespace paddle {
+namespace framework {
+
+struct RWLock {
```
Comment (typhoonzero) [inline paddle/fluid/framework/rw_lock.h:22]: Can add an `RWLockGuard` to reduce manual unlock.

File: paddle/fluid/framework/selected_rows.cc [hunk 4]
```diff
@@ -120,66 +120,76 @@ bool SelectedRows::HasKey(int64_t key) const {
...
-    const std::vector<int64_t>& keys, framework::Tensor* value) const {
+int64_t SelectedRows::AutoGrownIndex(int64_t key, bool auto_grown) {
+  rwlock_->RDLock();
+  auto iter = id_to_index_.find(key);
+  if (iter == id_to_index_.end()) {
+    rwlock_->UNLock();
+    if (!auto_grown) {
+      PADDLE_THROW("key %d not found", key);
+    }
+    rwlock_->WRLock();
```
Comment (typhoonzero) [inline paddle/fluid/framework/selected_rows.cc:131]: maybe `lock_guard.upgrade()`

**Review by seiriosPlus — APPROVED** [review]
LGTM！

File: paddle/fluid/framework/selected_rows.h [hunk 5]
```diff
@@ -127,9 +133,10 @@ class SelectedRows {
   // SelectedRows are simply concated when adding together. Until a
   // SelectedRows add a Tensor, will the duplicate rows be handled.
   Vector<int64_t> rows_;
+  std::unordered_map<int64_t, int64_t> id_to_index_;
```
Comment (typhoonzero) [inline paddle/fluid/framework/selected_rows.h:136]: Please keep in mind that adding a member variable here may increase construct time of `SelectedRows`. If this is only used for dist lookuptable, can use a sub class.

**Review by typhoonzero — APPROVED** [review]
LGTM with some advise.

File: paddle/fluid/framework/rw_lock.h (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/rw_lock.h:22]: good idea

File: paddle/fluid/framework/selected_rows.cc (hunk 4 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/selected_rows.cc:131]: ok

**Pull request merged**